### ISSUE TITLE
ERROR: object identifier out of order

### ISSUE THREAD

**Issue by zanfona734** [malonge/RagTag#94]
Hi Malonge,
Thank you for making RagTag available.  I'm wondering if you can shed some light on an error I'm seeing:

ERROR: object identifier out of order

I am using this command:
ragtag.py scaffold -u -w --mm2-params '-x asm20' reference.fasta target.fasta

I see from a different user's issue that the problem is caused by the program trying to write an improper agp file, but I'm not sure how to prevent/fix that.  If you are willing to check, I would be happy to share my raw data files.

**Comment by zanfona734** [comment]
Update:
I also tried omitting the minimap parameters:
**ragtag.py scaffold -u -w --mm2-params '-x asm20' reference.fasta target.fasta**

And the error remains, so it isn't that.

**Comment by malonge** [comment]
Hi there,

I would disregard #21. This shouldn't be influenced by the size of the alignment file. Could you provide the list of the reference and target assembly sequence headers?

Thanks

**Comment by malonge** [comment]
My first guess is that you have multiple sequences with the same header in your fasta files. Please check that and let me know.

**Comment by zanfona734** [comment]
I think I see:  The reference headers are unique but come directly from Genbank, where there is a space after the accession.  If only the first white-space delimited field is considered the header, it might be interpreted as non-unique.   I will test this and post the results here.

**Comment by malonge** [comment]
Yes the first field in the fasta header must uniquely identify each sequence. This is why the output AGP file was considered invalid, since the object ID depends on the fasta ID, and the object IDs must be unique for valid AGP files.

**Comment by zanfona734** [comment]
OK, no, that isn't it.  I renumbered both the reference and the header and the error persists.

You can see the renumbered target and reference here.  Note: the reference has been chopped into smaller fragments.

https://zenodo.org/record/5813768

**Comment by malonge** [comment]
Did you delete the old alignment file? Or else run with `-w`? If you are using the same alignment files (which is default), then RagTag would still be working with the old IDs.

**Comment by malonge** [comment]
If the alignment step was really expensive for you, I suppose you could try to manually change the alignment headers to match the new headers. But if the alignment was relatively cheap, then just generate new alignments.

**Comment by zanfona734** [comment]
The alignment is cheap.  I deleted the alignment file after each test , though.  And I also ran with -w.

**Comment by malonge** [comment]
And you still see the error?

**Comment by zanfona734** [comment]
Yes, still seeing the error.
Note: some runs on other datasets work fine.  I am not clear as to what is wrong with this one.

**Comment by malonge** [comment]
Hmm yeah I'll take a look at these fasta files and see if I can diagnose it. Sorry for the delay

**Comment by zanfona734** [comment]
"Sorry for the delay"   ?
Wow, you respond immediately, on a Sunday, for software you don't charge for.   I am very grateful for your help!

**Comment by malonge** [comment]
So it seems like a rather foolish bug on my part. The files you sent had fasta headers that were just incremented integers. So for example, both the reference and query assembly both have a sequence labeled as `1`. So if there is one scaffold corresponding to reference sequence `1` it will be labelled as `1_RagTag`. And if the query sequence `1` is unplaced and you use the `-u` option, that will also be named `1_RagTag`. Duplicate names are not allowed in AGP files so this causes the bug.

The immediate solution would be to either not use `-u` or to rename the headers so that there are no duplicates, even across both assemblies. So for example, you can name the reference headers as ref1, ref2, ..., refn and you can name query headers as query1, query2, ..., queryn. Let me know if this ends up working for you.

Even if that solves your problem, I will leave this issue open as a bug in case I want to address this formally in future releases.

**Comment by zanfona734** [comment]
Ah!  
Thank you very much for figuring that out.  That explains a lot, including why this worked sometimes but not others.  
Easy fix for me now.
Very grateful for your help with this.
